accepted/2021/architecture-targeting.md

```diff
@@ -0,0 +1,498 @@
+
+- Use the existing `any` RID to mean portable.
+- Expose a new CLI arg and MSBuild property for portable apps, like  `--portable` and `<Portable>true</Portable>`.
+
+Recommendation: Start with using just the RID. Wait for feedback to enable more polished experiences to validate that portable is an important scenario.
+
+There is also an existing `win` RID. It could be used in a similar way as `any` but limited to just Windows family RIDs. ASP.NET Core developers deploying and/or developing apps on some combination of Windows x86, x64, and Arm64 may find that scenario useful. We should not proactively develop that scenario, but wait for feedback.
+
+In addition, the SDK would no longer generate apphost EXEs for portable apps by default. There already is an opt-in model for EXEs (`<UseAppHost>true</UseAppHost>`). Portable app users can rely on that.
+
+The rationale for making apphost EXEs opt-in is that they are a non-portable asset, creating a strange asymmetry in an otherwise portable app. Perhaps, we should make it easy for developers to generate an apphost only for debug builds to aid the development experience. That said, if developers want a better development experience, they could just as easily build RID-specific apps (the proposed default) for development and then build a portable app with no EXE for production where an EXE is presumably not needed.
+
+Note: Portable apps are not a meaningful option for client apps since almost all client apps require an executable launcher, and there is only ever one generated. Portable apps are only meaningful for commandline apps.
+
+Note: In theory, we could consider an option to create multiple apphosts for different RIDs, either of different names or in different directories. Again, this capability would need to be opt-in.
+
+### Rationalize RID folders for RID-specific and portable apps
+
+There have been some issues filed (see appendix) that suggest that managed libraries are not resilient to RID-specific libraries changing location as a result of changing build type (portable to RID-specific). This would also be true for single file deployment builds.
+
+We could enable (as an option) creating `runtimes` folders for RID-specific apps to make the folder layout the same across deployment types. Doing that would put single file at a significant disadvantage. We'd rather encourage the ecosystem to support both folder layouts. If we had to choose one of these deployments types to disadvantage, it would be portable (with the `runtimes` folders), not RID-specific, which is the likely the opposite of what developers are tuning for today.
+
+We need to invest more in docs and tools to the end of clear and actionable guidance.
+
+### Use SDK RID as default RID
+
+RID-specific apps should default to the SDK RID as an implicit RID, and no longer require a RID to be specified. The SDK already does this for the apphost for portable apps. This proposal is a straightforward extension of that behavior.
```

Comment:
How is this going to work on configurations where SDK is built for the more generic RID than I'm on (e.g. some Linux flavours) ?

Reply:
Please also consider how NuGet lock files should work. Today, the following commands result in an error: https://github.com/NuGet/Home/issues/9195

```text
dotnet new console
dotnet restore --use-lock-file
dotnet restore --locked-mode -r linux-x64
```

While that looks like a contrived example (why would someone restore locked mode for a specific RID?), consider that customers want to use locked mode on CI build pipelines, and that `dotnet publish -r linux-x64` does an implicit restore. Bang, restore failed, so publishing failed, and CI failed.

Customers want to use lock files and locked mode so they can use floating versions, but still get consistent restores across all machines, and also some package substitution prevention, as lock files ensure the package's content hash is the same when packages are extracted.

There needs to be a way for the SDK to ensure that the set of packages NuGet is asked to restore (plus the set of RIDs that NuGet is informed about) is consistent, regardless of what RID is being built/published by the comment line arguments.

This spec seems to be a higher level than these details, but lock files need to be designed to be workable from the top. NuGet can't have a workable lock file experience when the SDK reports a different set of packages on different actions.

Reply:
For the most part, we don't offer RIDs for different Linux distros (AFAIK). The key outlier is Alpine. Outside of Alpine, I don't see that mattering. Today, the product only cares about glibc and musl c.